feat(pipelines): Expose stack output namespaces in custom `pipelines.Step`s

tobni opened the pull request:
Closes https://github.com/aws/aws-cdk/issues/23000 as per request from https://github.com/aws/aws-cdk/issues/23000#issuecomment-1324379670.

In order for custom steps to include `CfnOutput` in their action configurations, there needed to be a way to access and/or generate the output variable namespaces. `ShellStep` already had this capability.

This change generalizes the `StackOutputReference` usages by letting implementors of `Step` define the `StackOutputReference`s their step consumes, instead of hardwiring it to `ShellStep`.

To actually consume the references, the `ICodePipelineActionFactory` provides a `StackOutputsMap` that exposes a method to 
 render `StackOutputReference`s into their assigned CodePipeline variable names.

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation reviewed (changes requested):
The pull request linter has failed. See the aws-cdk-automation comment below for failure reasons. If you believe this pull request should receive an exemption, please comment and provide a justification.

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to an integration test file and the resulting snapshot.

PRs must pass status checks before we can provide a meaningful review.

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to an integration test file and the resulting snapshot.

PRs must pass status checks before we can provide a meaningful review.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 43bed9c699626cce17f26b9c411f9ad46db8d6f1
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ff7aa7a3-7816-4423-b0d3-c355364e9d60%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 9ec69ac3c3f1fe3d1c65e6748e19f15060f7cb80
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8f546bc7-14aa-4619-b8c9-9859b0e410c6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: fd37e4e7572df154422b0b7f0fc14a1c98228ace
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9c65f022-56a6-4819-9752-dc5fe42c38f9%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: cee8a46567d802ebc11feb7163151892ee157271
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=5ae98e06-3cd3-417b-9308-4b30c99ae6b1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

HBobertz reviewed (commented):
Looks like a pretty feature to me, and looks like it's working as intended. Feels like quite a lot of effort for not an immense payoff compared to the workaround. As I am new on the team I am going to run this by another engineer before approving

rix0rrr commented on packages/@aws-cdk/pipelines/README.md:
```diff
@@ -861,6 +861,43 @@ class MyJenkinsStep extends pipelines.Step implements pipelines.ICodePipelineAct
+      actionName: options.actionName,
+      runOrder: options.runOrder,
+      // Map the reference to the variable name the CDK has generated for you.
+      userParameters: {stackOutput: options.stackOutputsMap.mapOutputReference(this.stackOutputReference)},
```
This shouldn't compile, since `stackOutputsMap` is optional (should require a `?.` or `!.`), but I'm pretty sure our code translator would balk at that.

Since `ProduceActionOptions` is much more of an output struct than an input struct, you have my permission to make the field required and silence the backwards compatibility checker.

tobni replied:
I've added the compat error to https://github.com/aws/aws-cdk/pull/23110/commits/5bd7d2cdbb936bd975a25b1afd428ebca100aa1f, is that sufficient?

rix0rrr replied:
Yes, that is it exactly

rix0rrr commented on packages/@aws-cdk/pipelines/lib/codepipeline/codepipeline-action-factory.ts:
```diff
@@ -70,6 +71,19 @@ export interface ProduceActionOptions {
+   *
+   * @default - No stack output references to map
+   */
+  readonly stackOutputsMap?: StackOutputsMap
```
```suggestion
  readonly stackOutputsMap?: StackOutputsMap;
```

rix0rrr commented on packages/@aws-cdk/pipelines/lib/codepipeline/codepipeline.ts:
```diff
@@ -313,6 +314,7 @@ export class CodePipeline extends PipelineBase {
   private _myCxAsmRoot?: string;
   private readonly dockerCredentials: DockerCredential[];
   private readonly cachedFnSub = new CachedFnSub();
+  private stackOutputs: StackOutputsMap
```
```suggestion
  private stackOutputs: StackOutputsMap;
```

rix0rrr commented on packages/@aws-cdk/pipelines/lib/helpers-internal/pipeline-graph.ts:
```diff
@@ -274,11 +274,9 @@ export class PipelineGraph {
-        const stack = this.queries.producingStack(output);
-        this.stackOutputDependencies.get(stack).dependBy(node);
-      }
+    for (const output of step.stackOutputDependencies) {
```
❤️

rix0rrr reviewed (commented):
Love this change! I'd like you to change a couple of small things, but overall thanks for generalizing this feature. Approved if you address my comments.

(Please do write a description of the motivation, the change and significant decisions made in the PR body though! Linking to the issue is not sufficient to cover the last 2 parts.)

tobni commented:
> Love this change! I'd like you to change a couple of small things, but overall thanks for generalizing this feature. Approved if you address my comments.
> 
> (Please do write a description of the motivation, the change and significant decisions made in the PR body though! Linking to the issue is not sufficient to cover the last 2 parts.)

I've addressed the comments (unsure if the compat check is correctly handled) and updated the PR description now.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 8c30b188b8cf43a1aa934964f2dff47d3b852955
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3e06b35a-c9ee-4a2f-a331-7e34ecca75f6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: b238b269c73dd10f183131a7d259cfc77f3c7bf7
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4f04d57d-275d-4865-81ad-717550ba0643%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 7097022ae035a323ccbe50995916a65f21c670be
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=90e4253e-ec1b-4e0a-bd14-112f22bbeb52%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

tobni commented:
@rix0rrr I don't understand the Mergify error, is there some configuration missing in my fork, or is it permissions in this repo that is incorrect?

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 7668fdd16e9da11efa53085bc06c9426a2a1ea7c
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a997de50-2cf1-40f9-9d3d-70a257d55e97%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
This PR cannot be merged because it has conflicts. Please resolve them. The PR will be considered stale and closed if it remains in an unmergeable state.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 7097406e6b775280cf7f5946afb1719a3ffdbc67
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=5b05d762-3620-4663-b5cd-ba180274ca37%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 34eb6ce878daff83de2964cc2d410645745a9e01
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c8c7404f-6fe3-4fe1-954b-b3a7708a0125%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
